## Programming part of Homework 2 (Data Structures,Fall 2020)

## Name: 呂柏鴻
## Student ID Number: 107820016

### Programming problem 1
Consider the problem of generating the series $S$ of numbers in Problem 3. Please use Python to implement the pseudo-code you provide for solving the problem. You should provide the iterative and recursive functions named with `Dec_to_Bin_Ite()` and `Dec_to_Bin_Rec()` respectively. Then, perform a comparison on these two versions by measuring the execution time. Besides, we also consider the number of recursive calls made in the recursive approaches. Hence, you need to

1. implement the iterative approach as a function named as `Dec_to_Bin_Ite()`,

2. provide the recursive version for the approach named as `Dec_to_Bin_Rec()`, and

3. compare these two function with the same input in terms of running time and write what you have observed.

We will use an in-built python library `timeit` and the module function `timeit.timeit()` for measuring the running time respective. In you observation, please indicate the ways to measure the time and this will be provided in the template.

**Note:** For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.

In [1]:
def Dec_to_Bin_Ite(n):
    bin = ''
    while n!=0:
        if n%2 == 1:
            bin = '1' + bin
        elif n%2 == 0:
            bin = '0' + bin
        n//= 2
    return bin
        

print(Dec_to_Bin_Ite(10))
print(Dec_to_Bin_Ite(19))
print(Dec_to_Bin_Ite(63))

In [2]:
def Dec_to_Bin_Rec(n):
    bin = ''
    if n == 0:
        return ''
    elif n%2 == 1:
        return Dec_to_Bin_Rec(n//2) + '1'
    elif n%2 == 0:
        return Dec_to_Bin_Rec(n//2) + '0'

    
print(Dec_to_Bin_Rec(10))
print(Dec_to_Bin_Rec(19))
print(Dec_to_Bin_Rec(63))

1010
10011
111111


In [3]:
#---------------------------------------------------------timeit
import timeit as ti
# n=1050
n_list = [1524, 38769, 1863576, 33659435, 2**50]
#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

for n in n_list:

    tstart = ti.default_timer() #timer start 計時器開啟
    for i in range(10000):
        Dec_to_Bin_Ite(n)
    tend = ti.default_timer() #timer end計時器關閉
    t_ite=tend - tstart#開始與結束的差值

    tstart = ti.default_timer() #timer start 計時器開啟
    for i in range(10000):
        Dec_to_Bin_Rec(n)
    tend = ti.default_timer() #timer end計時器關閉
    t_rec=tend - tstart#開始與結束的差值

    print("Iterative on ", n)
    print("iterative approach:", t_ite)
    print("Recursive approach:", t_rec)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    Dec_to_Bin_Ite(n)
    
def recursive_test():
    Dec_to_Bin_Rec(n)

for n in n_list:
    
    t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
    t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)

    print("Iterative on ", n)
    print("iterative approach:", t_ite)
    print("Recursive approach:", t_rec)


Running time comparison (Timer Method 1):
Iterative on  1524
iterative approach: 0.01228639999999992
Recursive approach: 0.013897700000000013
Iterative on  38769
iterative approach: 0.018355800000000033
Recursive approach: 0.020976400000000006
Iterative on  1863576
iterative approach: 0.024627500000000024
Recursive approach: 0.026914899999999964
Iterative on  33659435
iterative approach: 0.02970000000000006
Recursive approach: 0.034427100000000044
Iterative on  1125899906842624
iterative approach: 0.07460109999999998
Recursive approach: 0.08374910000000002
Running time comparison (Timer Method 2):
Iterative on  1524
iterative approach: 0.01327299999999998
Recursive approach: 0.01410099999999992
Iterative on  38769
iterative approach: 0.01876219999999995
Recursive approach: 0.021433499999999883
Iterative on  1863576
iterative approach: 0.02470620000000001
Recursive approach: 0.027260799999999863
Iterative on  33659435
iterative approach: 0.03047349999999982
Recursive approach: 0.0352681

#### Please write down the observations you have after the comparisons:
##### Inputs (values of n) for five experiments

1. 1524

2. 38769

3. 1863576

4. 33659435

5. 1125899906842624

##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.013022899999999948|0.018870300000000007|0.025905200000000073|0.030445100000000003|0.0764281|
|Recursion      |0.013905899999999916|0.022622400000000042|0.027013799999999977|0.035066799999999954|0.08196159999999997|

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.013050100000000064|0.019255299999999975|0.024386100000000077|0.031243800000000155|0.0753855000000001|
|Recursion      |0.014386099999999846|0.021681199999999956|0.027779200000000115|0.03406379999999998|0.08325130000000014|

*******************************************************************************************************

##### Observations
After five experiments, we can find iteration is more faster than recursion and it makes more obviously when the nuber is bigger, but it has little difference.


### Programming problem 2
A ***permutation*** $P$ of size $n$ is a way of arranging the numbers between 1 and $n$, where each number must be used and can only be used once. For example, 5 4 1 2 3 is a permutation of 1 2 3 4 5 where $P(5)=1$, $P(4)=2$, $P(1)=3$, $P(2)=4$, and $P(3)=5$. An ***inversion*** in a permutation $P$ is a pair of numbers $i \mbox{ and} j \in\{1, \ldots, n\}$ with $i<j$ such that the larger number appears to the left of the smaller one in the permutation, that is $P(j)<P(i)$. In the example, the pair of (3, 5) is an inversion in $P$. The ***inversion number*** of a permutation is the total number of inversions. Please write the Python fuctions mentioned in Problem 4 and compare the performance in term of running time. The things to be provided are as below.

1. Implement the iterative approach using Python and name the function as `inversion_number_iterative()`.

2. Please implement the linear recursive version you provide in Problem 4(c) with the function name of `inversion_number_recursive()`.

3. Write the binary recursive version you provide in Problem 4(d) with the function named as `inversion_number_tworecurs()`.

4. Compare these functions with the same input in terms of running time and write what you have observed.

In [4]:
def inversion_number_iterative(P):
    inversion_number = 0
    for pivot in range(len(P)):
        for target in range(pivot + 1, len(P)):
            if P[pivot] > P[target]:
                inversion_number+=1
    return inversion_number



# perm=[4, 6, 2, 5, 1, 3]
# perm = [12, 11, 13, 5, 6, 7]
perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_iterative(perm))

The inversion number of  [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82] is 19


In [5]:
def inversion_number_recursive(P):
    if len(P) == 1:
        return 0
    else:
        return recursive(P, 0, len(P)-1)

def recursive(P, pivot, target):
    if pivot == target:
        return inversion_number_recursive(P[1:])
    elif P[pivot] > P[target]:
        return recursive(P, pivot, target - 1) + 1
    else:
        return recursive(P, pivot, target - 1)



# perm = [x for x in reversed(range(1,200))]
perm=[4, 6, 2, 5, 1, 3]
# perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_recursive(perm))

The inversion number of  [4, 6, 2, 5, 1, 3] is 10


In [6]:
def inversion_number_tworecurs(P):
    
    if len(P) < 2:
        return 0

    left, right, sorted_list = P[:len(P)//2], P[len(P)//2:], P
    it_l, it_r, it_sl = 0, 0, 0
    count = inversion_number_tworecurs(left) + inversion_number_tworecurs(right)

    while it_l < len(left) and it_r < len(right):
        count += it_r
        while it_r < len(right) and left[it_l] > right[it_r]:
            sorted_list[it_sl] = right[it_r]
            it_r, it_sl, count = it_r + 1, it_sl + 1, count + 1
        sorted_list[it_sl] = left[it_l]
        it_l, it_sl = it_l + 1, it_sl + 1

    while it_l < len(left):
        sorted_list[it_sl] = left[it_l]
        it_l, it_sl = it_l + 1, it_sl + 1
        count += it_r

    while it_r < len(right):
        sorted_list[it_sl] = right[it_r]
        it_r, it_sl = it_r + 1, it_sl + 1

    return count
    

# perm = [x for x in reversed(range(1,200))]
perm=[4, 6, 2, 5, 1, 3]
# perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_tworecurs(perm))

The inversion number of  [1, 2, 3, 4, 5, 6] is 10


In [7]:
#---------------------------------------------------------timeit
import timeit as ti

perm_list = [
    [323, 282, 804, 859, 245],
    [372, 28, 83, 779, 319, 64, 382, 84, 98, 435],
    [343, 950, 717, 768, 884, 948, 354, 309, 974, 104, 100, 118, 438, 193, 448, 304, 613, 871, 678, 386],
    [591, 75, 73, 238, 965, 650, 779, 278, 707, 643, 777, 912, 949, 39, 40, 90, 382, 55, 650, 769, 372, 460, 26, 927, 351, 434, 370, 987, 922, 37, 936, 612, 552, 331, 847],
    [802, 755, 609, 487, 883, 621, 791, 700, 678, 327, 120, 91, 375, 29, 171, 588, 153, 381, 851, 985, 903, 532, 574, 703, 372, 873, 926, 125, 531, 261, 299, 880, 562, 829, 964, 469, 109, 148, 333, 916, 751, 419, 298, 344, 895, 548, 853, 982, 507, 690]
]
# perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]

# ---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

for perm in perm_list:
    
    print("Permutation:", perm)

    tstart = ti.default_timer() #timer start 計時器開啟

    for i in range(10000):
        inversion_number_iterative(perm)
    tend = ti.default_timer() #timer end計時器關閉
    t_ite=tend - tstart#開始與結束的差值

    tstart = ti.default_timer() #timer start 計時器開啟
    for i in range(10000):
        inversion_number_recursive(perm)
    tend = ti.default_timer() #timer end計時器關閉
    t_rec=tend - tstart#開始與結束的差值

    tstart = ti.default_timer() #timer start 計時器開啟
    for i in range(10000):
        inversion_number_tworecurs(perm)
    tend = ti.default_timer() #timer end計時器關閉
    t_twr=tend - tstart#開始與結束的差值

    print('The inversion number of is', inversion_number_iterative(perm),'(iterat)')
    print('The inversion number of is', inversion_number_recursive(perm),'(linear)')
    print('The inversion number of is', inversion_number_tworecurs(perm),'(binary)')

    print("Iterative approach:", t_ite)
    print("Recursive approach:", t_rec)
    print("Binrecurs approach:", t_twr)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    inversion_number_iterative(perm)

def recursive_test():
    inversion_number_recursive(perm)

def binrecurs_test():
    inversion_number_tworecurs(perm)

for perm in perm_list:
    
    print("Permutation:", perm)
    
    t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
    t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)
    t_twr = ti.timeit("binrecurs_test()", setup="from __main__ import binrecurs_test", number = 10000)

    print("Iterative approach:", t_ite)
    print("Recursive approach:", t_rec)
    print("Binrecurs approach:", t_twr)

Running time comparison (Timer Method 1):
Permutation: [323, 282, 804, 859, 245]
The inversion number of is 0 (iterat)
The inversion number of is 0 (linear)
The inversion number of is 0 (binary)
Iterative approach: 0.017681000000000058
Recursive approach: 0.02164150000000009
Binrecurs approach: 0.03714510000000004
Permutation: [372, 28, 83, 779, 319, 64, 382, 84, 98, 435]
The inversion number of is 0 (iterat)
The inversion number of is 0 (linear)
The inversion number of is 0 (binary)
Iterative approach: 0.042445499999999914
Recursive approach: 0.07286260000000011
Binrecurs approach: 0.09157820000000005
Permutation: [343, 950, 717, 768, 884, 948, 354, 309, 974, 104, 100, 118, 438, 193, 448, 304, 613, 871, 678, 386]
The inversion number of is 0 (iterat)
The inversion number of is 0 (linear)
The inversion number of is 0 (binary)
Iterative approach: 0.1278127
Recursive approach: 0.2734198000000001
Binrecurs approach: 0.20999329999999983
Permutation: [591, 75, 73, 238, 965, 650, 779, 278, 7

#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments

1. [323, 282, 804, 859, 245]

2. [372, 28, 83, 779, 319, 64, 382, 84, 98, 435]

3. [343, 950, 717, 768, 884, 948, 354, 309, 974, 104, 100, 118, 438, 193, 448, 304, 613, 871, 678, 386]

4. [591, 75, 73, 238, 965, 650, 779, 278, 707, 643, 777, 912, 949, 39, 40, 90, 382, 55, 650, 769, 372, 460, 26, 927, 351, 434, 370, 987, 922, 37, 936, 612, 552, 331, 847]

5. [802, 755, 609, 487, 883, 621, 791, 700, 678, 327, 120, 91, 375, 29, 171, 588, 153, 381, 851, 985, 903, 532, 574, 703, 372, 873, 926, 125, 531, 261, 299, 880, 562, 829, 964, 469, 109, 148, 333, 916, 751, 419, 298, 344, 895, 548, 853, 982, 507, 690]

##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.017982599999982085|0.04265989999998965|0.12628140000003896|0.33142510000004677|0.6320180000000164|
|Recursion      |0.021263200000021243|0.07291299999997136|0.27343229999996765|0.8707490000000462|1.8336553000000322|
|Binrecurs      |0.03730380000001787|0.0896940000000086|0.20646770000007564|0.39466860000004544|0.5926742999999988|

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.017037200000004304|0.04206650000003265|0.12514650000002803|0.3278497999999672|0.6388945000001058|
|Recursion      |0.020470600000066952|0.0729562999999871|0.27029589999995096|0.8748986000000514|1.7838435000001027|
|Binrecurs      |0.03618910000000142|0.08881250000001728|0.21437620000006063|0.4017339000000675|0.5974969999999757|

*******************************************************************************************************

##### Observations

In these experiments, we can find the process time is binary recursion < iteration < linear recursion  when the list is big, and in the case of small list, the process time is iteration < linear recursion < binary recursion.
